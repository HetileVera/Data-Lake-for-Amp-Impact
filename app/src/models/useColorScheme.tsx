import { useTheme } from "@chakra-ui/react";

export function useColorScheme() {
  const colorScheme = useTheme().name;
  const tone = (value: number | undefined) => (value ? `${colorScheme}.${value}` : colorScheme);

  return { tone, colorScheme };
}
