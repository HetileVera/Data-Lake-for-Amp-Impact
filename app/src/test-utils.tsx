import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";

import { theme } from "themes/orange";

export const CustomChakraProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <ChakraProvider theme={theme}>
      <div data-testid="test-root">{children}</div>
    </ChakraProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "wrapper">) => render(ui, { wrapper: CustomChakraProvider, ...options });

export * from "@testing-library/react";

export { customRender as render };
