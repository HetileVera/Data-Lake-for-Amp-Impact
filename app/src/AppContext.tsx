import React, { createContext, useContext } from "react";
import isEmpty from "lodash/isEmpty";
import throttle from "lodash/throttle";

import { AppStore, IAppStore } from "./models/AppStore";
import { IInstallation } from "./models/Installation";
import { isLocalDev, websiteUrlOrigin, assetWebsiteUrl, secretKey } from "./helpers/settings";
import { onSnapshot } from "mobx-state-tree";

// Create an instance of the root store (a.k.a appStore)
const appStore = AppStore.create({});

// Store state to local storage
const storeLocally = throttle(() => appStore.saveLocal(), 5000);

// Store state in S3 when applicable
const storeRemotely = throttle(() => appStore.saveRemote(), 30000);

// The initialization function for the root store
export function initializeStore() {
  if (!isEmpty(secretKey)) console.warn("Access key is bundled!");

  if (isLocalDev) {
    console.log(`isLocalDev '${isLocalDev}'`);
    console.log(`websiteUrlOrigin '${websiteUrlOrigin}'`);
    console.log(`assetWebsiteUrl '${assetWebsiteUrl}'`);

    if (window.location.host !== "localhost:3000") {
      console.warn("You should run the app on localhost!");

      throw new Error("App should run on localhost for development mode!");
    }
  }

  appStore.loadLocal();
  onSnapshot(appStore, (snapshot) => {
    storeLocally();
    storeRemotely();
  });
}

// Create the store context
export const StoreContext = createContext<IAppStore | undefined>(undefined);

// Create the store provider component
export const StoreProvider: React.FunctionComponent = ({ children }) => {
  return <StoreContext.Provider value={appStore}>{children}</StoreContext.Provider>;
};

// useStore hook
export function useStore(): IAppStore {
  const store = useContext(StoreContext);

  if (!store) {
    throw new Error("useStore() must be used within a StoreProvider");
  }

  return store;
}

// useInstallation hook
export function useInstallation(): IInstallation {
  const store = useStore();
  const installation = store.installation;

  if (!installation) {
    throw new Error("An installation instance was not created yet");
  }

  return installation;
}
