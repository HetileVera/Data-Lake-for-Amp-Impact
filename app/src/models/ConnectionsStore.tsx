import { useEffect } from "react";
import isEmpty from "lodash/isEmpty";
import { types, Instance, getParent } from "mobx-state-tree";

import { useInstallation } from "AppContext";
import { listAppflowConnections } from "api/list-appflow-connections";

import { BaseStore, isStoreNew, isStoreError, isStoreReady, isStoreLoading, isStoreReLoading } from "./BaseStore";
import { ICredentials } from "./helpers/Credentials";
import { IInstallation } from "./Installation";

/**
 * Represents the store for the AppFlow Salesforce connection names
 */
export const ConnectionsStore = BaseStore.named("ConnectionsStore")
  .props({
    connectionNames: types.array(types.string),
  })

  .views((self) => ({
    get empty() {
      return isEmpty(self.connectionNames);
    },

    get credentials(): ICredentials {
      const parent: IInstallation = getParent(self);
      return parent.credentials;
    },

    get region(): string {
      const parent: IInstallation = getParent(self);
      return parent.region;
    },
  }))

  .actions((self) => {
    const superReset = self.reset;
    return {
      reset() {
        superReset();
        self.connectionNames.clear();
      },

      async doLoad() {
        const accessKey = self.credentials.accessKey;
        const secretKey = self.credentials.secretKey;
        const region = self.region;

        const rawConnectionNames = await listAppflowConnections({ accessKey, secretKey, region });
        self.runInAction(() => {
          self.connectionNames.replace(rawConnectionNames);
        });
      },
    };
  });

// see https://mobx-state-tree.js.org/tips/typescript
export interface IConnectionsStore extends Instance<typeof ConnectionsStore> {}

export function useConnectionsStore() {
  const installation = useInstallation();
  const store = installation.connectionsStore;

  useEffect(() => {
    if (!isStoreNew(store)) return;
    store.load();
  }, [store]);

  return {
    isError: isStoreError(store),
    isReady: isStoreReady(store),
    isLoading: isStoreLoading(store),
    isReloading: isStoreReLoading(store),
    store,
  };
}
