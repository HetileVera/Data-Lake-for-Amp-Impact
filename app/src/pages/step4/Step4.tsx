import React from "react";
import { Box, Heading, Text, HStack, Button } from "@chakra-ui/react";
import { ArrowForwardIcon, ArrowBackIcon } from "@chakra-ui/icons";
import { observer } from "mobx-react";
import { useNavigate } from "react-router-dom";

import { useInstallation } from "AppContext";
import { Layout } from "components/Layout";
import { DataModelTablePreview } from "pages/step4/data-model-table-preview/DataModelTablePreview";

import { ReviewGrid } from "pages/step4/ReviewGrid";

export const Step4 = observer(() => {
  const installation = useInstallation();
  const step = installation.reviewStep;
  const navigate = useNavigate();

  const handlePrevious = () => {
    navigate("/steps/3");
  };

  const handleSubmit: React.MouseEventHandler<HTMLButtonElement> = async () => {
    step.markCompleted();
    navigate("/steps/5");
  };

  React.useEffect(() => {
    step.markStarted();
    window.scrollTo(0, 0);
  }, [step]);

  return (
    <Layout
      step={4}
      title="Review and Confirm"
      description="Did we get everything right? Take a look at all the answers you provided and confirm your selections."
    >
      <Box borderRadius="lg" boxShadow="base" bg="pink.50" mt={4} p={8} pb={10} pt={10}>
        <ReviewGrid />
        <DataModelTablePreview colorScheme="pink" showCaption={false} />
      </Box>

      <Box borderRadius="lg" boxShadow="base" bg="pink.50" mt={4} p={8} pb={10} pt={10}>
        <Heading size="md" pt="0px" pb="10px" color="pink.600" letterSpacing="-1px">
          What's next?
        </Heading>

        <Text>
          We will start the process of setting up the data lake at AWS and importing your Amp Impact and other selected Salesforce object records. Depending on the number of records to
          import, the next step might take 15 minutes to a few hours to complete.
        </Text>
      </Box>

      <HStack justifyContent="space-between" p={3} pt={6} mb={12}>
        <Button colorScheme="pink" size="md" _hover={{ bg: "pink.100" }} leftIcon={<ArrowBackIcon />} variant="ghost" onClick={handlePrevious}>
          Previous
        </Button>
        <Button
          id="step4-btn-next"
          colorScheme="pink"
          size="md"
          loadingText="Processing"
          rightIcon={<ArrowForwardIcon />}
          ml={3}
          onClick={handleSubmit}
        >
          Next
        </Button>
      </HStack>
    </Layout>
  );
});
