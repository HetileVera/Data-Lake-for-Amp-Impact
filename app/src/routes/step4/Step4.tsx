import React from "react";
import { Box, Container, Heading, Text, HStack, Button, ChakraProvider, Stack } from "@chakra-ui/react";
import { ArrowForwardIcon, ArrowBackIcon } from "@chakra-ui/icons";
import { observer } from "mobx-react";
import { useNavigate } from "react-router-dom";

import { useInstallation } from "AppContext";
import { theme } from "themes/pink";
import { Header } from "components/Header";
import { CurvedBox } from "components/CurvedBox";
import { StepsIndicator } from "components/StepsIndicator";
import { DataModelTablePreview } from "routes/step4/DataModelTablePreview";

import { ReviewGrid } from "routes/step4/ReviewGrid";

export const Step4 = observer(() => {
  const installation = useInstallation();
  const step = installation.reviewStep;
  const navigate = useNavigate();

  const handlePrevious = () => {
    navigate("/steps/3");
  };

  const handleSubmit: React.MouseEventHandler<HTMLButtonElement> = async () => {
    step.markCompleted();
    navigate("/steps/5");
  };

  React.useEffect(() => {
    step.markStarted();
    window.scrollTo(0, 0);
  }, [step]);

  return (
    <ChakraProvider theme={theme}>
      <CurvedBox bgGradient={theme.gradients.bgLight} />

      <Header />

      <Box position="relative">
        <CurvedBox />
        <Box bg="pink.600" w="full" position="relative">
          <Container maxW="container.md" pt="15px" pb="15px">
            <StepsIndicator current={4} />
          </Container>
        </Box>

        <Container maxW="container.md" pt="16px" position="relative">
          <Box float="right" fontSize="small" textAlign="center" color="pink.100">
            <Text>Installation ID</Text>
            <Text>{installation.id}</Text>
          </Box>

          <Heading display="inline-block" size="lg" pt="0px" pb="10px" color="white" letterSpacing="-1px">
            Review and confirm
          </Heading>

          <Box color="purple.50" mb={10}>
            <Stack direction="row" spacing="0" align="left" justifyContent="space-between">
              <Box mb={0}>Did we get everything right? Take a look at all the answers you provided and confirm your selections.</Box>
            </Stack>
          </Box>

          <Box borderRadius="lg" boxShadow="base" bg="pink.50" mt={4} p={8} pb={10} pt={10}>
            <ReviewGrid />
            <DataModelTablePreview colorScheme="pink" showCaption={false} />
          </Box>

          <Box borderRadius="lg" boxShadow="base" bg="pink.50" mt={4} p={8} pb={10} pt={10}>
            <Heading size="md" pt="0px" pb="10px" color="pink.600" letterSpacing="-1px">
              What's next?
            </Heading>

            <Text>
              We will start the process of setting up the datalake at AWS and importing your Salesforce records. Depending on the number of records to
              import, the next step might take 15 minutes to a few hours to complete.
            </Text>
          </Box>

          <HStack justifyContent="space-between" p={3} pt={6} mb={12}>
            <Button colorScheme="pink" size="md" _hover={{ bg: "pink.100" }} leftIcon={<ArrowBackIcon />} variant="ghost" onClick={handlePrevious}>
              Previous
            </Button>
            <Button colorScheme="pink" size="md" loadingText="Processing" rightIcon={<ArrowForwardIcon />} ml={3} onClick={handleSubmit}>
              Next
            </Button>
          </HStack>
        </Container>
      </Box>
    </ChakraProvider>
  );
});
