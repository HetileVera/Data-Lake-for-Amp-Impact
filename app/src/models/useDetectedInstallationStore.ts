import { isStoreError, isStoreLoading, isStoreReady, isStoreReLoading } from "models/BaseStore";

import { DetectedInstallationStore } from "./DetectedInstallationsStore";

// Create an instance of this store
const store = DetectedInstallationStore.create({});

export function useDetectedInstallationStore() {
  return {
    isError: isStoreError(store),
    isReady: isStoreReady(store),
    isLoading: isStoreLoading(store),
    isReloading: isStoreReLoading(store),
    store,
  };
}
