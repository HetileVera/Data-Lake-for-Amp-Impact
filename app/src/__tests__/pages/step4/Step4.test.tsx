import { cleanup, screen } from "@testing-library/react";
import { create } from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";

import { render, CustomChakraProvider } from "test-utils";

import * as appContext from "AppContext";
import * as metadataStore from "models/MetadataStore";
import { Step4 } from "pages/step4/Step4";

const spyScrollTo = jest.fn();

describe("Step4 component", () => {
  beforeEach(() => {
    Object.defineProperty(global.window, "scrollTo", { value: spyScrollTo });
  });

  afterEach(() => {
    spyScrollTo.mockClear();
    cleanup;
  });

  test("snapshot", () => {
    // @ts-ignore
    jest.spyOn(appContext, "useInstallation").mockImplementation(() => {
      return {
        id: "1",
        startDate: "2022-01-01",
        startedBy: "test-user",
        accountId: "test-account-id",
        region: "us-east-1",
        appFlowConnectionName: "test-appflow-connection-name",
        credentials: [],
        deploymentStep: "1",
        nextStepNumber: "2",
        importOptionsStep: {
          markStarted: jest.fn(),
          markCompleted: jest.fn(),
        },
        reviewStep: {
          markStarted: jest.fn(),
          markCompleted: jest.fn(),
        },
      };
    });
    // @ts-ignore
    jest.spyOn(metadataStore, "useMetadataStore").mockImplementation(() => {
      return {
        isReady: true,
        store: {
          selectedObjects: [{}],
        },
      };
    });

    const tree = create(
      <CustomChakraProvider>
        <appContext.StoreProvider>
          <BrowserRouter>
            <Step4 />
          </BrowserRouter>
        </appContext.StoreProvider>
      </CustomChakraProvider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("render properly", () => {
    // @ts-ignore
    jest.spyOn(appContext, "useInstallation").mockImplementation(() => {
      return {
        id: "1",
        startDate: "2022-01-01",
        startedBy: "test-user",
        accountId: "test-account-id",
        region: "us-east-1",
        appFlowConnectionName: "test-appflow-connection-name",
        credentials: [],
        deploymentStep: "1",
        nextStepNumber: "2",
        importOptionsStep: {
          markStarted: jest.fn(),
          markCompleted: jest.fn(),
        },
        reviewStep: {
          markStarted: jest.fn(),
          markCompleted: jest.fn(),
        },
      };
    });
    // @ts-ignore
    jest.spyOn(metadataStore, "useMetadataStore").mockImplementation(() => {
      return {
        isReady: true,
        store: {
          selectedObjects: [{}],
        },
      };
    });

    render(
      <appContext.StoreProvider>
        <BrowserRouter>
          <Step4 />
        </BrowserRouter>
      </appContext.StoreProvider>
    );

    expect(screen.getAllByText(/review and confirm/i)).toBeTruthy();
    expect(screen.getAllByText(/region/i)).toBeTruthy();
    expect(screen.getByText(/did we get everything right/i)).toBeInTheDocument();
    expect(screen.getByText("Previous")).toBeInTheDocument();
    expect(screen.getByText("Next")).toBeInTheDocument();
  });
});
