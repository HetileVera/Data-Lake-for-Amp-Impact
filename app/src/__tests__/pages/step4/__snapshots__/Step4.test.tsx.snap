// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step4 component snapshot 1`] = `
Array [
  <div
    data-testid="test-root"
  >
    <div
      className="css-0"
      id="step4"
    >
      <div
        className="css-1v55i00"
        data-testid="curved-box"
      />
      <div
        className="chakra-container css-1szj4qk"
      >
        <div
          className="chakra-stack css-olw0lh"
        >
          <div
            className="chakra-stack css-18lyn6q"
          >
            <div
              className="css-1i3fafl"
            >
              <img
                className="chakra-image css-0"
                src="Salesforce_Corporate_Logo_RGB.png"
              />
            </div>
            <div
              className="chakra-stack__divider css-r2vkku"
            />
            <div
              className="css-nutmff"
            >
              <img
                className="chakra-image css-ir5tdx"
                src="AWS_logo_RGB_BLK.png"
              />
            </div>
          </div>
          <div
            className="chakra-stack css-1n7gsjr"
          >
            <div
              className="css-18n4g2t"
            >
              <h2
                className="chakra-heading css-1opsxio"
                id="h2-title"
              >
                Data Lake for Nonprofits, Powered by AWS
              </h2>
            </div>
          </div>
        </div>
      </div>
      <div
        className="css-79elbk"
      >
        <div
          className="css-t7j6xp"
          data-testid="curved-box"
        />
        <div
          className="css-4e82ru"
        >
          <div
            className="chakra-container css-h66qep"
          >
            <div
              className="chakra-stack css-84zodg"
            >
              <span
                className="css-9xsoz7"
              >
                1
              </span>
              <span
                className="css-9xsoz7"
              >
                2
              </span>
              <span
                className="css-9xsoz7"
              >
                3
              </span>
              <span
                className="css-hfqsm"
              >
                4
              </span>
              <span
                className="css-ulmpab"
              >
                5
              </span>
              <span
                className="css-ulmpab"
              >
                6
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        className="chakra-container css-11yz6e2"
      >
        <h2
          className="chakra-heading css-noy8un"
        >
          Review and Confirm
        </h2>
        <div
          className="css-1kveix6"
        >
          <p
            className="chakra-text css-0"
          >
            Region: 
            us-east-1
          </p>
          <p
            className="chakra-text css-0"
          >
            Installation ID: 
            1
          </p>
        </div>
        <div
          className="css-vfbt1y"
        >
          <div
            className="chakra-stack css-1s6236i"
          >
            <div
              className="css-1y0bycm"
            >
              <div
                className="css-0"
              >
                Did we get everything right? Take a look at all the answers you provided and confirm your selections.
              </div>
            </div>
          </div>
        </div>
        <div
          className="css-1ikeo3e"
        >
          <div
            className="css-9kpwwf"
          >
            <div
              className="css-1dlsxld"
            >
              AWS Region
            </div>
            <div
              className="css-hl0cvm"
            >
              US East (N. Virginia)
            </div>
            <div
              className="css-1dlsxld"
            >
              Connection Name
            </div>
            <div
              className="css-hl0cvm"
            >
              test-appflow-connection-name
            </div>
            <div
              className="css-1dlsxld"
            >
              Import Schedule
            </div>
            <div
              className="css-hl0cvm"
            />
            <div
              className="css-1dlsxld"
            >
              Data Model
            </div>
            <div
              className="css-hl0cvm"
            >
              We will import 
              1
               objects
            </div>
          </div>
          <table
            className="chakra-table css-94ssnp"
            role="table"
          >
            <thead
              className="css-0"
            >
              <tr
                className="css-0"
                role="row"
              >
                <th
                  className="css-111khnu"
                >
                  Type
                </th>
                <th
                  className="css-111khnu"
                  data-is-numeric={true}
                >
                  # Fields
                </th>
              </tr>
            </thead>
            <tbody
              className="css-0"
            >
              <tr
                className="css-4g6ai3"
                onClick={[Function]}
                role="row"
              >
                <td
                  className="css-1d5g9w7"
                  role="gridcell"
                >
                  <div
                    className="chakra-stack css-snbr68"
                  >
                    <svg
                      className="chakra-icon css-m6c8f6"
                      focusable={false}
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                        fill="currentColor"
                      />
                    </svg>
                    <div
                      className="css-0"
                    >
                       Â 
                      <p
                        className="chakra-text css-1fk7407"
                      >
                        ( 
                         )
                      </p>
                    </div>
                  </div>
                </td>
                <td
                  className="css-1d5g9w7"
                  data-is-numeric={true}
                  role="gridcell"
                >
                  N/A
                  /
                  N/A
                </td>
              </tr>
            </tbody>
            <tfoot
              className="css-5wov5z"
            >
              <tr
                className="css-0"
                role="row"
              >
                <th
                  className="css-111khnu"
                />
                <th
                  className="css-111khnu"
                  data-is-numeric={true}
                >
                  N/A
                </th>
              </tr>
            </tfoot>
          </table>
        </div>
        <div
          className="css-1ikeo3e"
        >
          <h2
            className="chakra-heading css-1t5ljy0"
          >
            What's next?
          </h2>
          <p
            className="chakra-text css-0"
          >
            We will start the process of setting up the data lake at AWS and importing your Salesforce records. Depending on the number of records to import, the next step might take 15 minutes to a few hours to complete.
          </p>
        </div>
        <div
          className="chakra-stack css-1aoc1cb"
        >
          <button
            className="chakra-button css-1112rcq"
            onClick={[Function]}
            type="button"
          >
            <span
              className="chakra-button__icon css-1wh2kri"
            >
              <svg
                aria-hidden={true}
                className="chakra-icon css-onkibi"
                focusable={false}
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
                  fill="currentColor"
                />
              </svg>
            </span>
            Previous
          </button>
          <button
            className="chakra-button css-1yog7ax"
            id="step4-btn-next"
            onClick={[Function]}
            type="button"
          >
            Next
            <span
              className="chakra-button__icon css-1hzyiq5"
            >
              <svg
                aria-hidden={true}
                className="chakra-icon css-onkibi"
                focusable={false}
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>,
  <span />,
]
`;
